src/coreclr/jit/hwintrinsic.cpp

```diff
@@ -58,6 +58,302 @@ const HWIntrinsicInfo& HWIntrinsicInfo::lookup(NamedIntrinsic id)
     return hwIntrinsicInfoArray[id - NI_HW_INTRINSIC_START - 1];
 }
 
+#if defined(TARGET_XARCH)
+const TernaryLogicInfo& TernaryLogicInfo::lookup(uint8_t control)
+{
+    // This table is 768 bytes and is about as small as we can make it.
+    //
+    // The way the constants work is we have three keys:
+    // * A: 0xF0
+    // * B: 0xCC
+    // * C: 0xAA
+    //
+    // To compute the correct control byte, you simply perform the corresponding operation on these keys. So, if you
+    // wanted to do (A & B) ^ C, you would compute (0xF0 & 0xCC) ^ 0xAA or 0x6A.
+    //
+    // This table allows us to compute the inverse information, that is given a control, what are the operations it
+    // performs. This allows us to determine things like what operands are actually used (so we can correctly compute
+    // isRMW) and what operations are performed and in what order (such that we can do constant folding in the future).
+    //
+    // The total set of operations supported are:
+    // * true:   AllBitsSet
+    // * false:  Zero
+    // * not:    ~value
+    // * and:    left & right
+    // * nand:   ~(left & right)
+    // * or:     left | right
+    // * nor:    ~(left | right)
+    // * xor:    left ^ right
+    // * xnor:   ~(left ^ right)
+    // * cndsel: a ? b : c; aka (B & A) | (C & ~A)
+    // * major:  0 if two+ input bits are 0
+    // * minor:  1 if two+ input bits are 0
+
+    // clang-format off
+    static const TernaryLogicInfo ternaryLogicFlags[256] = {
+        /* FALSE */           { TernaryLogicOperKind::False,  TernaryLogicUseFlags::None, TernaryLogicOperKind::None,   TernaryLogicUseFlags::None, TernaryLogicOperKind::None,   TernaryLogicUseFlags::None },
+        /* norABC */          { TernaryLogicOperKind::Nor,    TernaryLogicUseFlags::ABC,  TernaryLogicOperKind::None,   TernaryLogicUseFlags::None, TernaryLogicOperKind::None,   TernaryLogicUseFlags::None },
```

Comment:
how does this work? When I read `norABC` I see that as `A nor B nor C` and having `TernaryLogicUseFlags::ABC` as `op1Use` while all the others are left `None` is slightly confusing to me.

Reply:
`norABC` is `~(A | B | C)`. They shorthand the ones that do the same operation on all three, where-as the full name would've been something like `norAorBC`

Most of these control bytes do exactly 2 operations at once and they are always read as `oper Input1 Input2`. So `andAorBC` is `And(A, Or(B, C))` (aka. `A & (B | C)`).

The exception is the `conditional select` instructions which do 3 operations (with the third one always being the condition code that "picks" bits from the first or second operation).